### ISSUE TITLE
Validators with low forging ratio should be dropped

### ISSUE THREAD

**Issue by cryptozero** [semuxproject/semux#98]
Someone with enough voting power could introduce multiple stale validators and disrupt the network. Also an attack on several nodes, abandoned or shutdown validators could affect transaction processed by the network.

Validators with low performance should be droped from the top of validator list. After n blocks validators will be whitelisted again, if they fail again to perform, drop time doubles.

**Comment by semuxproject** [comment]
The validator slashing mechanism needs more research and formal proof, as it introduces more state and may be exploited validators.

**Comment by cryptozero** [comment]
One idea to improve health of network and reward best delegates:

- Less performing validator get dropped from list.
- Next delegate enter the list to forge.
- Dropped validator can enter the list again after one round is completed (100 blocks) in place of the less performing validator.
- If reentered validator is the dropped again in less than n blocks, drop time doubles.

This solution should incorporate some protection against DDoS targeting specific nodes.

**Comment by AlexAltertrader** [comment]
I think you need a operation in the BAN in blockchain, issued by the validators working for the dead validator. After that, this validator is dropped from validators list for a ban period.

**Comment by phash** [comment]
instant ban might not be the best solution... second chance before ban maybe?
It can always happen, a validator has just one bad luck block but is reliable all the other time. Look at the 90-99% validators. They are not bad validators, they just missed one block in a while..

So: second chance for x blocks - if the x blocks are over, the flag is removed. If the validator misses a second block, he gets a ban for y blocks and drops out of validators for that time span

**Comment by AlexAltertrader** [comment]
@phash I agree, the second chance is necessary. And maybe the third. But no more, skipping the third round means that the validator has problems.

**Comment by orogvany** [comment]
So in this case, what constituted a failed round? 100% failed forgings in the round?  Remember that voting is as important as forging, so I'd probably extend that to 'no votes/forgings in round', basically if the node is down.

Sadly, we do not currently keep track of voting information history in the blockchain I don't think.

**Comment by AlexSemuxTrader** [comment]
I think that validators with a poor percentage of processed blocks should be discarded from the validators.
The maximum score window for processed blocks should be 300 blocks.
Then validators can be validators that have the next percent of the processed blocks.
For the first 10 blocks, -50%
For the first 100 blocks-80%
For the first 200 blocks-90%
For the first 300 blocks-95%
